src/quarot/hadamard_utils.py

```diff
@@ -0,0 +1,4201 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+# Based on the implementation in https://github.com/spcl/QuaRot/
+
+import math
+
+import fast_hadamard_transform
+import torch
+
+
+def get_hadK(n, transpose=False):
+    hadK, K = None, None
+    if n % 172 == 0:  # llama-2-7b up
+        assert (is_pow2(n // 172))
+        K = 172
+        hadK = get_had172().T if transpose else get_had172()
+    elif n % 156 == 0:  # llama-1-30b 3x hidden
+        assert (is_pow2(n // 156))
+        K = 156
+        hadK = get_had156().T if transpose else get_had156()
+    elif n % 140 == 0:  # llama-1-30b intermediate 
+        assert (is_pow2(n // 140))
+        K = 140
+        hadK = get_had140().T if transpose else get_had140()
+    elif n % 108 == 0:  # llama-1-13b intermediate 
+        assert (is_pow2(n // 108))
+        K = 108
+        hadK = get_had108().T if transpose else get_had108()
+    elif n % 60 == 0:  # llama-1-13b 3x hidden
+        assert (is_pow2(n // 60))
+        K = 60
+        hadK = get_had60().T if transpose else get_had60()
+    elif n % 52 == 0:  # llama-1-13b 1x hidden
+        assert (is_pow2(n // 52))
+        K = 52
+        hadK = get_had52().T if transpose else get_had52()
+    elif n % 36 == 0:
+        assert (is_pow2(n // 36))
+        K = 36
+        hadK = get_had36().T if transpose else get_had36()
+    elif n % 28 == 0:
+        assert (is_pow2(n // 28))
+        K = 28
+        hadK = get_had28().T if transpose else get_had28()
+    elif n % 40 == 0:
+        assert (is_pow2(n // 40))
+        K = 40
+        hadK = get_had40().T if transpose else get_had40()
+    elif n % 20 == 0:
+        assert (is_pow2(n // 20))
+        K = 20
+        hadK = get_had20().T if transpose else get_had20()
+    elif n % 12 == 0:
+        assert (is_pow2(n // 12))
+        K = 12
+        hadK = get_had12().T if transpose else get_had12()
+    else:
+        assert (is_pow2(n))
+        K = 1
+
+    return hadK, K
+
+
+def matmul_hadU(X, transpose=False):
```

Comment:
@sashkboos could you please explain what `matmul_hadU` is doing? and what `K` and `hadK` are?

Reply:
@nailimixaM 

This is supposed to do the hadamard transformation (with arbitrary sizes). This was from QuiP repo and I think we do not use this function in the code anymore.

Reply:
@sashkboos We use it when we want to [calculate the random hadamard Q](https://github.com/microsoft/TransformerCompression/blob/88e3ce602106562d88a50ef4a04f9b682986049f/src/quarot/hadamard_utils.py#L99 ) used in the whole network

Reply:
Ah yes, you are right. 

this is simple recursive calculation of Hadamard transformation based on the kronecker product. `hadK` is the hadamard matrix for the cases where you do not have power-of-two matrices (the supported cases are in `get_hadK` function and are saved in `K`).